### ISSUE TITLE
Incorporating Sinusoidal pitching motion in body.

### ISSUE THREAD

**Issue by Akarshr1234** [aharwood2/LUMA#17]
Hi,
In my project I need to incorporate sinusoidal pitching of any airfoil. I was stuck in that. I have also seen issue #5 but I don't know what to upgrade. So, I was asking if it is possible to do so with LUMA and if yes how can we do so.

Hope you could help.
Thank  You.

**Comment by ianhinder** [comment]
As you saw in #5, this isn't implemented yet, but it should be possible to get it working in a rudimentary way with a bit of coding.  I am assuming that you want to use an IBM body, not a bounceback body.  In that case, assuming that the IBM body is rigid, you should be able to add code to set the position and velocity of each node at each timestep.  LUMA should then use this updated velocity to calculate the correct force to apply back to the fluid.

I think you want to write a new function, say, ObjectManager::ibm_apply_prescribed_movement, and call it in ObjectManager::ibm_moveBodies in ObjectManager_ops_ibm.cpp between the blocks "Loop through flexible bodies and apply FEM" and "Update IBM markers".  

This new function would have to select which bodies (in `iBody`) to apply the movement to.  I would implement the actual movement routine in the IBBody class, which is the one that hold the markers.   In the IBBody object, you will need to loop over `markers` from `Body`, the superclass, and update `position` (from Marker, the superclass) and markerVel (from IBMarker).  I would start by applying a constant velocity and updated position to all markers in the x direction, and simulating something like an impulsively started flat plate (e.g. a filament in 2D), and testing your result against the literature.  If this works, then you can think about how to deal with the rotation.  You will obviously need to specify a centre of rotation and a time-varying rotation, from which you can calculate the position and (angular) velocity of each node.

**Comment by paulakash24** [comment]
Hi, 
We have incorporated the function `ObjectManager::ibm_apply_prescribed_movement` as suggested. The function takes current time instant as input and updates the position and velocity of the markers. The function is able to update the position (correctly) and velocity of the marker. But the velocity values in the domain start becoming nan as the motion starts and it propagates from the region near the filament to the whole domain.

 We were printing the marker velocity values at certain positions of the filament during the motion, and they seemed alright.

Currently, we are trying to give pitching motion to a rigid filament (pitching point is the leading edge (1st marker)).

We just want to know, if we are missing anything while updating the velocities of the markers. We are looping through each body and inside that looping through each marker and assigning the velocity as:

```
iBody[ib].markers[i].markerVel[0] = new_u;
iBody[ib].markers[i].markerVel[1] = new_v;
```
where `new_u` and `new_v` are the  calculated velocities of each marker at the current timestep for the pitching motion.


We are attaching two snapshots at time instant 20 and 40:

**20 ---->**
![1](https://user-images.githubusercontent.com/42206087/123581359-56eaf780-d7f9-11eb-9919-e4ae0789cc52.jpeg)


**40 ---->**
![2](https://user-images.githubusercontent.com/42206087/123581362-594d5180-d7f9-11eb-956f-d1fc07652dd2.jpeg)

**Comment by Akarshr1234** [comment]
Hi,
We had made the desired function for moving the body. The function is basically iterating through all `iBody` present in the simulation and it is updating the position and velocity markers of them.
We are using  a single body (Rigid Filament Array) to perform sinusoidal pitching. We are successfully changed the position the body at every time instant. We had also updated the Velocity marker at every time instant. But we are getting NaN values for output fluid velocity and its density (Ux, Uy, Rho).
This is the function we used.
```
void ObjectManager::ibm_apply_prescribed_movement(int timeinstant){

double actual_time = timeinstant/2000.0; // dimensional time
double omega = 8; // angular frequency in radian per second
double amplitude = 10; // in degrees
double amp_radian = amplitude * (3.14159265359/180); // amplitude in radian
double instant_angle = (amp_radian * sin(omega * actual_time)); // theta(t) = theta_max*sin(2*pi*f*t)

for (size_t ib = 0; ib < iBody.size(); ib++) { // looping through each body

	// setting the pivot position: postion of the 1st marker
	double x_pivot = iBody[ib].markers[0].position0[0];
	double y_pivot = iBody[ib].markers[0].position0[1];

	for (size_t i = 0; i < iBody[ib].markers.size(); i++) { // looping through each marker

		#if (L_DIMS == 2)
			// getting the x,y co-ordinates if its 2D
			double x_pos = iBody[ib].markers[i].position0[0];
			double y_pos = iBody[ib].markers[i].position0[1];
		#endif

		// calculating the new position of a marker by rotating the point about the pivot pitching point
		double new_x_pos = (x_pos-x_pivot)*cos(instant_angle) - (y_pos-y_pivot)*sin(instant_angle) + x_pivot;
		double new_y_pos = (x_pos-x_pivot)*sin(instant_angle) + (y_pos-y_pivot)*cos(instant_angle) + y_pivot;

		// updating marker position
		iBody[ib].markers[i].position[0] = new_x_pos;
		iBody[ib].markers[i].position[1] = new_y_pos;

		// calculating r
		double r = sqrt(pow((new_x_pos - x_pivot),2) + pow((new_y_pos - y_pivot),2));

		// calculating u and v of a marker
		double new_u = -amp_radian*r*omega*sin(instant_angle)*cos(omega*actual_time);
		double new_v = +amp_radian*r*omega*cos(instant_angle)*cos(omega*actual_time);

		//updating marker velocities
		iBody[ib].markers[i].markerVel[0] = new_u;
		iBody[ib].markers[i].markerVel[1] = new_v;
		}
	}
}
```

Do we have to update any other variable?? Please tell us why we are getting NaN values.

Thank You.

**Comment by ianhinder** [comment]
Could it be something to do with whether the bodies are marked as "flexible" in LUMA?  Have you had to make other changes to the code in order to get ibm_moveBodies to be called?  The reason I ask is that ibm_moveBodies is only called if ObjectManager::hasFlexibleBodies is true.  It might be that if you have made the body flexible, then other parts of LUMA might think that they have to do evolve the structure as a flexible body, and this might be causing problems.  Can you describe exactly what you did?  e.g. could you fork the repository and push all your changes to a branch in the fork, so we can see them?

**Comment by paulakash24** [comment]
Yes, you are right. We removed the condition before calling the ibm_moveBodies. Now it gets called irrespective of whether the body is rigid or flexible.

We did make another small change/correction recently.

Anyway we doing the fork repository as you have suggested.

Thanks

**Comment by paulakash24** [comment]
Hi I have forked the repo.. Kindly check..

**Comment by ianhinder** [comment]
I noticed that you commented out ibm_subIterate; was that intentional?

In your commit (https://github.com/paulakash24/LUMA/commit/f31b9197c17befe5880b6f1d636d21aa39d66018), you seem to have changed a lot of whitespace, and there are some changes in there that don't look relevant, so it's very hard to see what you were intending to change.

I can make a guess as to what is wrong, but you would have to try to track down whether this is the case yourself.  LUMA identifies bodies as either flexible or not flexible.  Currently, only flexible bodies can move, and when they do move, certain things need to be updated, like the IBM support points, epsilon, etc.  It might be that this is not happening because the body is not flexible.  

For example, this code
```
	// Loop through flexible bodies and update the support points for all valid markers existing on this rank
	for (size_t ib = 0; ib < iBody.size(); ib++) {

		// Only do if on this grid level
		if (iBody[ib]._Owner->level == level && iBody[ib].isFlexible)
			ibm_findSupport(static_cast<int>(ib));
	}
```
in ObjectManager_ops_ibm.cpp:86, is only refinding the support points if the body is flexible.  I suspect you could remove that `&& iBody[ib].isFlexible` part and have it recompute the support points for all bodies.  Maybe that would fix it?  But in general, you could search through the code and look for things that are only done for flexible bodies, and see if they should be done for moving bodies as well.

**Comment by paulakash24** [comment]
Yes. After doing the fork, I figured these things out actually (what you just said). I removed those flexibility check for calling ibm_findsupport in ObjectManager_ops_ibm.cpp and all such checks in ObjectManager_ops_ibm_mpi::ibm_updateMarkers.
I also used the conversion factor of velocities now.

And sorry if I changed the code layout a lot.

Now it is consistent. I can mention the body as rigid. And give rigid body motion.

I commented subIterate as I thought it was not necessary as it is a rigid body motion (positions and velocities are calculated). So currently, we are in a state that "rigid body motion" is working fine for all kind of bodies. This part is done probably. We need to validate.

Now we are trying to give different kinds of motion to flexible bodies. And here we require some suggestions. We have two things in mind:

1. Give pitching motion to a flexible filament (about the leading edge).
2. We want to prescribe motion (motion of any king, i.e., plunging, or some other where we will specify the position and velocity to the first marker) to the leading edge of the flexible filament. And we would expect the rest of the flexible body to follow that first point.

Can we get some suggestions on these? It would be very helpful. This would enable us to approach a large set of problems which we were thinking of tackling.

Thanks
